### PR TITLE
Intake

### PR DESCRIPTION
**Pull request by omegaladon**
Creates intake and pivot subsystems.

Simulation and tests need work

### CODE REVIEW

File: src/main/java/frc/robot/commands/gpm/IntakeNote.java
Comment (Pennedrose) [inline src/main/java/frc/robot/commands/gpm/IntakeNote.java:1]: forever lmao. i think this shouldn't be pushed into main until fleshed out and tested

File: src/main/java/frc/robot/subsystems/Drivetrain.java
Comment (Pennedrose) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:1]: why r u deleting so much logging stuff?

Comment (omegaladon) [reply src/main/java/frc/robot/subsystems/Drivetrain.java:1]: dunno. i didnt touch drivetrain. maybe i did something bad with github or smth lol

File: src/main/java/frc/robot/commands/gpm/IntakeNote.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    public void initialize() {
+        intake.setMode(Intake.Mode.INTAKE);
+    }
+
+    // TODO: This code might need some adjusting, as the cone passes through the intake.
+    @Override
+    public boolean isFinished() {
+        return stallDebouncer.calculate(
+                intake.getCurrent() >= IntakeConstants.INTAKE_CURRENT_STOP
+        );
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/commands/gpm/IntakeNote.java:30]: This command probably needs the indexer (should be merged soon) as well as the intake. Because it's continuous, we don't want to stop it when it has a higher current or a note in it. It also might not have a high current long enough to trigger the debouncer. Using the indexer's hasNote() would be better.

File: src/main/java/frc/robot/constants/IntakeConstants.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    public static final int CONTINUOUS_CURRENT_LIMIT = 25;
+    public static final int PEAK_CURRENT_LIMIT = 55;
+    public static final double PEAK_CURRENT_DURATION = 0.1;
+    public static final boolean ENABLE_CURRENT_LIMIT = true;
+
+    public static final double INTAKE_STALL_TIME = 0.2;
+    public static final double INTAKE_CURRENT_STOP = 10;
+
+    private IntakeConstants() {
+    }
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/constants/IntakeConstants.java:22]: Remove the unused constructor.

File: src/main/java/frc/robot/subsystems/gpm/Intake.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import frc.robot.constants.IntakeConstants;
+
+
+public class Intake extends SubsystemBase {
+
+    public enum Mode {
+        INTAKE(IntakeConstants.INTAKE_POWER), DISABLED(0)
+        ;
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/subsystems/gpm/Intake.java:17]: Put this ; on the previous line. You can put the 2 modes on separate lines if you want to, but you can choose.

File: src/main/java/frc/robot/subsystems/gpm/Intake.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            return power;
+        }
+    }
+
+    private final CANSparkFlex motor;
+    private final DigitalInput sensor;
+
+    private int countSim = 0;
+    private DIOSim IntakeSensorDioSim;
+    private  boolean foo = true;
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/subsystems/gpm/Intake.java:35]: Rename "foo" to something more descriptive.

File: src/main/java/frc/robot/subsystems/gpm/Intake.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    private void publish() {
+        SmartDashboard.putBoolean("Intake Sensor", sensor.get());
+    }
+
+    public void setMode(Mode mode) {
+        this.mode = mode;
+    }
+
+    public boolean hasNote() {
+        return sensor.get();
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/subsystems/gpm/Intake.java:66]: On the indexer, this is inverted. If this is the same type of sensor, should it be inverted here?

File: src/main/java/frc/robot/subsystems/gpm/Intake.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+        if (!sensor.get()) {
+            System.out.println("Object detected");
+
+        } else if (sensor.get()) {
+            System.out.println("No object detected");
+
+        } else {
+            System.out.println("=P");
+        }
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/subsystems/gpm/Intake.java:83]: Don't print in periodic. Remove this.

File: src/main/java/frc/robot/subsystems/gpm/Intake.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    @Override
+    public void simulationPeriodic() {
+
+        //change values every 1/2 secound
+         if (countSim++ > 25) {
+            countSim = 0;
+
+            IntakeSensorDioSim.setValue(foo);
+
+            foo = ! foo;
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/subsystems/gpm/Intake.java:99]: Make the simulation better. It shouldn't just toggle the sensor every half second. Maybe add a number showing the speed? You can probably use a flywheel sim for this.

File: vendordeps/Phoenix5.json
Comment (Kyle-Eldridge) [inline vendordeps/Phoenix5.json:1]: Delete this. We're using Phoenix 6.

**Review by Kyle-Eldridge — CHANGES_REQUESTED** [review]

**Comment by Ash1R** [comment]
Needs to have Neo motor for centering, motor spins clockwise. Just use some constant speed.

File: src/main/java/frc/robot/RobotContainer.java [hunk 8]
```diff
@@ -4,12 +4,15 @@
...
 import com.pathplanner.lib.auto.AutoBuilder;
 import edu.wpi.first.wpilibj.DriverStation;
+import edu.wpi.first.wpilibj.PS5Controller;
+import edu.wpi.first.wpilibj.XboxController;
 import edu.wpi.first.wpilibj.livewindow.LiveWindow;
 import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
 import edu.wpi.first.wpilibj2.command.Command;
 import edu.wpi.first.wpilibj2.command.InstantCommand;
+import edu.wpi.first.wpilibj2.command.InstantCommand;
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/RobotContainer.java:14]: Why are there duplicate and unused imports? Delete these.

File: src/main/java/frc/robot/RobotContainer.java [hunk 9]
```diff
@@ -52,15 +57,17 @@ public RobotContainer(RobotId robotId) {
...
-        SmartDashboard.putData("Intake", new IntakeNote(intake, index));
         SmartDashboard.putData("shoot", new InstantCommand(() -> shooter.setTargetRPM(1500)));
         SmartDashboard.putData("shoot off", new InstantCommand(() -> shooter.setTargetRPM(0)));
         break;
 
       case TestBed2:
+        // gc = new XboxController(0);
+        intake = new Intake(); //gc
+        // SmartDashboard.putData("Rumble on", new InstantCommand(() -> intake.noteRumble(1)));
+        // SmartDashboard.putData("Rumble off", new InstantCommand(() -> intake.noteRumble(0)));
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/RobotContainer.java:70]: Remove commented out code.

File: src/main/java/frc/robot/controls/GameControllerDriverConfig.java [hunk 10]
```diff
@@ -19,6 +22,7 @@
  */
 public class GameControllerDriverConfig extends BaseDriverConfig {  
   private final GameController kDriver = new GameController(Constants.DRIVER_JOY);
+  private final XboxController xbox = new XboxController(5);
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/controls/GameControllerDriverConfig.java:25]: This should not be here. This class is for the driver's controller only; put the controller for subsystems in Operator.java, which doesn't exist yet.This should be in a separate PR.

File: src/main/java/frc/robot/controls/XboxDriverConfig.java [hunk 11]
```diff
@@ -0,0 +1,5 @@
+package frc.robot.controls;
+
+public class XboxDriverConfig {
+    
+}
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/controls/XboxDriverConfig.java:5]: If you are creating this class, put something in it. You can copy and paste most things from GameControllerDriverConfig. Also, I think the game controllers might be xbox controllers.

File: src/main/java/frc/robot/subsystems/gpm/Intake.java [hunk 12]
```diff
@@ -1,7 +1,8 @@
 package frc.robot.subsystems.gpm;
 
-import com.revrobotics.CANSparkFlex;
 import com.revrobotics.CANSparkLowLevel;
+import com.revrobotics.CANSparkMax;
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/subsystems/gpm/Intake.java:4]: I thought this was a neo vortex with a SparkFlex controller.

File: src/main/java/frc/robot/subsystems/gpm/Intake.java [hunk 13]
```diff
@@ -27,20 +28,24 @@ public double getPower() {
         }
     }
 
-    private final CANSparkFlex motor;
+    public final CANSparkMax motor;
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/subsystems/gpm/Intake.java:31]: Instance variables should be private.

File: src/main/java/frc/robot/subsystems/gpm/Intake.java [hunk 14]
```diff
@@ -27,20 +28,24 @@ public double getPower() {
...
     private Mode mode;
 
 
-    public Intake() {
-        motor = new CANSparkFlex(IntakeConstants.MOTOR_ID, CANSparkLowLevel.MotorType.kBrushless);
+    public Intake() { //XboxController gc
+        motor = new CANSparkMax(IntakeConstants.MOTOR_ID, CANSparkLowLevel.MotorType.kBrushless);
         sensor = new DigitalInput(IntakeConstants.SENSOR_ID);
-        mode = Mode.DISABLED;
+        mode = Mode.INTAKE;
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/subsystems/gpm/Intake.java:46]: When you deploy, it should be default disabled. Also, remove all commented out references to a controller.

File: src/main/java/frc/robot/subsystems/gpm/Intake.java [hunk 15]
```diff
@@ -68,19 +73,7 @@ public boolean hasNote() {
 
     @Override
     public void periodic() {
-        motor.set(mode.getPower());
-
```
Comment (Kyle-Eldridge) [inline src/main/java/frc/robot/subsystems/gpm/Intake.java]: This should still be here. We need to set the motor.

**Review by Kyle-Eldridge — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 12 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java:4]: yes

File: src/main/java/frc/robot/commands/DefaultDriveCommand.java [hunk 16]
```diff
@@ -43,7 +43,7 @@ public void execute() {
         sideTranslation *= slowFactor;
         rotation *= driver.getIsSlowMode() ? DriveConstants.kSlowRotFactor : 1;
 
-        int allianceReversal = DriverStation.getAlliance().get() == Alliance.Red ? 1 : -1;
+        int allianceReversal = DriverStation.getAlliance().get() == Alliance.Blue ? 1 : -1;
```
[the remark this answers is not in the document]
Comment (Kyle-Eldridge) [reply src/main/java/frc/robot/commands/DefaultDriveCommand.java:44]: Change this back.

**Comment by Kyle-Eldridge** [comment]
@Ash1R Don't mark the conversations as resolved until you make the changes in them or reply with why they don't need to be made. If you made these changes on your computer, push them so I can see them here.

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 15 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java]: done

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 14 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java]: done

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 13 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java]: done

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 12 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java:4]: done

File: src/main/java/frc/robot/controls/XboxDriverConfig.java (hunk 11 above)
Comment (Ash1R) [reply src/main/java/frc/robot/controls/XboxDriverConfig.java]: done

File: src/main/java/frc/robot/controls/GameControllerDriverConfig.java (hunk 10 above)
Comment (Ash1R) [reply src/main/java/frc/robot/controls/GameControllerDriverConfig.java]: done

File: src/main/java/frc/robot/RobotContainer.java (hunk 9 above)
Comment (Ash1R) [reply src/main/java/frc/robot/RobotContainer.java]: done

File: src/main/java/frc/robot/RobotContainer.java (hunk 8 above)
Comment (Ash1R) [reply src/main/java/frc/robot/RobotContainer.java]: done

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 6 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java]: done

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 5 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java]: done

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 4 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java]: done

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 3 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java]: done

File: src/main/java/frc/robot/constants/IntakeConstants.java (hunk 2 above)
Comment (Ash1R) [reply src/main/java/frc/robot/constants/IntakeConstants.java]: done

File: src/main/java/frc/robot/commands/gpm/IntakeNote.java (hunk 1 above)
Comment (Ash1R) [reply src/main/java/frc/robot/commands/gpm/IntakeNote.java]: done

File: src/main/java/frc/robot/commands/gpm/IntakeNote.java [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+        this.intake = intake;
+        this.stallDebouncer = new Debouncer(IntakeConstants.INTAKE_STALL_TIME, Debouncer.DebounceType.kBoth);
+    }
+
+    @Override
+    public void initialize() {
+        intake.setMode(Intake.Mode.INTAKE);
+    }
+
+    // TODO: This code might need some adjusting, as the cone passes through the intake.
```
[the remark this answers is not in the document]
Comment (Ash1R) [reply src/main/java/frc/robot/commands/gpm/IntakeNote.java]: done

File: src/main/java/frc/robot/commands/DefaultDriveCommand.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (Ash1R) [reply src/main/java/frc/robot/commands/DefaultDriveCommand.java]: done

File: src/main/java/frc/robot/subsystems/gpm/Intake.java (hunk 7 above)
Comment (Ash1R) [reply src/main/java/frc/robot/subsystems/gpm/Intake.java]: done by Arnav

**Review by Ash1R — APPROVED** [review]
kyle will probably find an issue with this that I don't see :), but looks good to me

**Review by Kyle-Eldridge — APPROVED** [review]
